### ISSUE TITLE
components made up of inline elements should preserve whitespace

### ISSUE THREAD

**Issue by ruslandoga** [phoenixframework/phoenix_live_view#3567]
👋 

### Environment

* Elixir version (elixir -v): 1.17.3
* Phoenix version (mix deps): 1.7.18
* Phoenix LiveView version (mix deps): 1.0.0
* Operating system: Darwin 24.1.0
* Browsers you attempted to reproduce this bug on (the more the merrier): 
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no:

### Actual behavior

Right now the HTML formatter treats all components as block elements, even if they consist of only inline elements.

This causes this component

```elixir
def styled_link(assigns) do
  ~H"""
  <a href={@href} style="color:red;">{render_slot(@inner_block)}</a>
  """
end
```

in this template

```heex
<p>
  <.styled_link href="some very long url and it goes on and on some very long url and it goes on and on  some very long url and it goes on and on ">go home</.styled_link>.
</p>
```

to be formatted like this

```heex
<p>
  <.styled_link href="some very long url and it goes on and on some very long url and it goes on and on  some very long url and it goes on and on ">
    go home
  </.styled_link>.
</p>
```

which changes how it gets rendered on page (it adds extra whitespace before the period).

Whereas the `a` tag by itself is not forced to "line-break"

```heex
<p>
  <a href="some very long url and it goes on and on some very long url and it goes on and on  some very long url and it goes on and on ">go home</a>.
</p>
```

^^^ this doesn't get re-formatted.

---

NOTE: I am aware of the `phx-no-format` workaround :)

---

### Expected behavior

Formatting of components should have the same rules as normal elements. If a component consists of only inline elements, then it should be treated as an inline element itself.

**Comment by josevalim** [comment]
Honestly, I don't know what to do here. We could implement this, but then it means we have even fewer opportunities to format the user code. Perhaps one option is to make it a configuration option in the formatter? Something like:

    inline_heex_elements: ~w(.styled_link foo-bar)

**Comment by josevalim** [comment]
I think we could try to respect the user preference. If there are no spaces after `>` then we will keep it as such. This way we don't need to know what is an inline element and what isn't, we just check how it is used. What do you think @feliperenan?

**Comment by kyrias** [comment]
I like that suggestion a lot since it means that it's very easy to specify on a case-by-case basis if you don't like how it turned out, though I'd personally still prefer for inline elements to be formatted as I suggested above because leaving them completely as-is ends up rather hard to read IMO.

It would probably need to be based on both the opening and closing tag though rather than just the opening tag.

**Comment by josevalim** [comment]
I sent a PR for this here: #3791

You can look at the failing tests and see how there are many situations where we would format before, such as `<section><div><h1>Hello</h1></div></section>`, but we cannot format now because we need spaces around it to be sure. 

Furthermore, note it does not make `<.link>` and friends behave fully as an inline element either. As per above, an inline element has to preserve spaces, as in `<b>Foo: </b>`, which the PR does not do. If we added this, then it means the formatter would almost be a no-op in many cases.

I think the best option here is to turn `link` and `button` into matchers. We will treat any component name with link or button in the name as inline element. This should at least make those surprising cases less common.

**Comment by SteffenDE** [comment]
What do other HTML formatters do?

> I think the best option here is to turn link and button into matchers. We will treat any component name with link or button in the name as inline element. This should at least make those surprising cases less common.

I feel like it should rather be something one annotates a component with? And we already have

https://github.com/phoenixframework/phoenix_live_view/blob/1dda262391f2bc4e3e5adde17112e45029171501/lib/phoenix_live_view/html_formatter.ex#L217

Is that relevant here?

**Comment by josevalim** [comment]
> I feel like it should rather be something one annotates a component with? And we already have

Right, my proposal was that, instead of it matching exactly the `link` component, it should match any component with the word `link` on it. We should probably include it both for `link` and `button`.

**Comment by SteffenDE** [comment]
Right, I understood the proposal, but I think relying on the name is a bit brittle?

I‘d think something like

```elixir
attr …

@type :inline

def my_button(assigns) do
  …
end
```

would be much nicer.

But with #3791 not sure if we need that?

**Comment by josevalim** [comment]
The formatter cannot rely on code being loaded or compiled, that would make it too slow and make it so you only run after compilation.

And #3791 only partially solves the problem at the cost of reducing the formatter ability to format code and it doesn't fully behave as an inline tag.

**Comment by SteffenDE** [comment]
That makes a lot of sense, I didn’t think about the code not being compiled… what about an option in the formatter config?

**Comment by josevalim** [comment]
Yes, we can definitely make it an option too. I think we should make it an option and but also make it relaxed, because no one wants to update it for every type of link or button they add.

**Comment by josevalim** [comment]
Closing in favor of the PR!

**Comment by Techn1x** [comment]
I'm a bit late to get my thoughts in here - here's a comment I made on a related issue that is fairly relevant, going into some of Prettier's default settings and why they might be so, and how we can learn from them for the liveview formatter.

Kinda thinking that trying to figure out if elements/components are inline or block is a misguided approach when it comes to template formatting.

https://github.com/phoenixframework/phoenix_live_view/issues/3718#issuecomment-2979253584

the TLDR

> As for how to treat whitespace regarding components? Since we can't really know correctly whether a component is inline or not, they should always be treated as inline. Anything else is guessing and can cause problems
